### PR TITLE
Soft debugger invocations refactoring

### PR DESCRIPTION
**Pull request by nerzhulart**
New IInvocableMethodOwnerMirror interface to avoid duplicating of Invoke*, BeginInvoke*, EndInvoke* methods in invocable mirrors. All the API is now static methods taking IInvocableMethodOwnerMirror as the 1st parameter. This refactoring allows to avoid multiple 'if (obj is *) else' in SoftDebuggerAdaptor.

This PR already merged in MonoDevelop/debugger-libs
https://github.com/mono/debugger-libs/pull/84

### CODE REVIEW

**Comment by vargaz** [comment]
Doesn't this change public api ?

**Comment by nerzhulart** [comment]
Hi, this keeps the API as was, but changes ABI, because some methods were made extensions. More info in a discussion  by URL above

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft-net_4_x.csproj [hunk 1]
```diff
@@ -1,178 +1,180 @@
...
  </ItemGroup>
-  <!-- To modify your build process, add your task inside one of the targets below and uncomment it. 
-       Other similar extension points exist, see Microsoft.Common.targets.
-  <Target Name="BeforeBuild">
-  </Target>
-  <Target Name="AfterBuild">
-  </Target>
-  -->
-  <PropertyGroup>
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (ludovic-henry) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft-net_4_x.csproj]: Please reduce whitespace noise to a minimum

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft-net_4_x.csproj [hunk 2]
```diff
@@ -13,7 +13,7 @@
     <IntermediateOutputPath>obj-net_4_x</IntermediateOutputPath>
     <GenerateTargetFrameworkAttribute>false</GenerateTargetFrameworkAttribute>
     <NoStdLib>True</NoStdLib>
-    
+
```
Comment (ludovic-henry) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft-net_4_x.csproj]: Please reduce whitespace noise to a minimum

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft-net_4_x.csproj (hunk 1 above)
Comment (nerzhulart) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft-net_4_x.csproj]: Yep, already doing this

**Comment by vargaz** [comment]
This doesn't compile because Mono.Debugger.Soft.dll.sources is not updated.

**Comment by vargaz** [comment]
Also, it does break api by changing the signatures of public methods.

**Comment by nerzhulart** [comment]
@ludovic-henry I've fixed whitespaces

**Comment by vargaz** [comment]
It still doesn't compile on osx/linux.

**Comment by nerzhulart** [comment]
The discussion about API change in debugger-libs https://github.com/mono/debugger-libs/pull/84/files/962fb01ca4f74a07ea0b52f31ab403c3c1a42b64#diff-4c7d45fc32f9891f0bef7b3344f0bec4R159

**Comment by vargaz** [comment]
Mono.Debugger.Soft is not part of xs, its part of mono, it needs to keep backward api compatibility since there could be other users beside xs.

**Comment by nerzhulart** [comment]
@vargaz If you try to save a backward compat you keep about 8-10 copy-pasted methods which causes low code quality. I was needed to make a simple change in this logic and had to copy-paste the change among all of these methods in every class. I've just fixed this to avoid code duplicating. For library customers nothing will be changed because the new methods are extensions to classes and lives in the same namespace, this means that library client just have to recompile his code.

**Comment by vargaz** [comment]
It should be possible to keep the api the same by adding internal methods and have the public methods call them, or adding new public methods in addition to the old ones.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft-net_4_x.csproj [hunk 3]
```diff
@@ -13,7 +13,7 @@
     <IntermediateOutputPath>obj-net_4_x</IntermediateOutputPath>
     <GenerateTargetFrameworkAttribute>false</GenerateTargetFrameworkAttribute>
     <NoStdLib>True</NoStdLib>
-
```
Comment (marek-safar) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft-net_4_x.csproj]: This file is auto generated please don't change it

Comment (nerzhulart) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft-net_4_x.csproj]: @marek-safar How to run autogen for this file?

**Comment by nerzhulart** [comment]
@vargaz if to keep the instance methods as was (and extract them to interface) when you need to add the new type like ObjectMirror you will have to implement all of this methods (9 items) in the new type, but all of these methods doesn't access to class fields/private properties/other implementation details. They only use Vm public property of IMirror, so I believe they are just static methods by their semantic.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft-net_4_x.csproj (hunk 3 above)
Comment (nerzhulart) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft-net_4_x.csproj]: @marek-safar So should I revert the changes in csproj?

Comment (marek-safar) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft-net_4_x.csproj]: Yes, it'll be overwritten anyway

**Comment by steveisok** [comment]
@vargaz should this PR be closed?

**Comment by vargaz** [comment]
Think so.

**Comment by steveisok** [comment]
Closing this ticket per our [old pull request policy](https://github.com/mono/mono/wiki/Pull-Request-Guide#old-pull-request-policy)
